File: packages/nodx-py/src/nodx/package_diagnostics.py
"""Structured package-reader diagnostic.

Mirrors `crates/nodx-package/src/lib.rs::PackageDiagnostic` so Python callers
can branch on `NODX-Exxx` codes instead of regexing English strings. The
default code is `NODX-E012` (resource limit / malformed envelope), matching
the Rust `diag()` helper. Severity follows `severity_for`.
"""


class NodxDiagnosticError(Exception):
    def __init__(self, code, message, severity=None):
        if severity is None:
            severity = severity_for(code)
        super().__init__(f"{code}: {message}")
        self.code = code
        self.severity = severity
        self.diagnostic_message = message

    def to_diagnostic(self):
        return {
            "code": self.code,
            "severity": self.severity,
            "message": self.diagnostic_message,
            "line": None,
            "column": None,
            "target": None,
        }


def severity_for(code):
    if code in ("NODX-E010", "NODX-E021"):
        return "error"
    if code == "NODX-E019":
        return "fatal"
    return "fatal"


def raise_diag(code, message):
    raise NodxDiagnosticError(code, message)

File: packages/nodx-js/src/packageDiagnostics.mjs
// Structured package-reader diagnostic.
//
// The Rust reference implementation surfaces every package-level failure as
// a `PackageDiagnostic { code, severity, message }`. This module mirrors that
// shape so JavaScript callers can branch on the code (NODX-E010, E012, E021)
// instead of pattern-matching English strings — which is what the parity
// audit flagged as the largest gap.
//
// Default code is NODX-E012 (resource limit / malformed envelope), matching
// the Rust `diag()` helper. Severity is derived from the code in the same way
// as `crates/nodx-package/src/lib.rs::severity_for`.

export class NodxDiagnosticError extends Error {
  constructor({ code, severity, message }) {
    super(`${code}: ${message}`);
    this.name = "NodxDiagnosticError";
    this.code = code;
    this.severity = severity;
    this.diagnosticMessage = message;
  }

  toDiagnostic() {
    return {
      code: this.code,
      severity: this.severity,
      message: this.diagnosticMessage,
      line: null,
      column: null,
      target: null,
    };
  }
}

export function severityFor(code) {
  switch (code) {
    case "NODX-E010":
    case "NODX-E021":
      return "error";
    case "NODX-E019":
      return "fatal";
    default:
      return "fatal";
  }
}

export function throwDiag(code, message) {
  throw new NodxDiagnosticError({ code, severity: severityFor(code), message });
}
